TAP-17: Remove signature wrapper from TUF spec (#138)
theupdateframework/taps: tap17.md

* TAP: 17
* Title: Remove Signature Wrapper from the TUF Specification
* Version: 1

> * Last-Modified: 11/11/2021

Author: hah, no date format confusions!

enables adopters to make their own decisions for their implementations, and
provides a mechanism for them to document their decisions.
[POUF-1](/POUFS/reference-POUF/pouf1.md) is the POUF for the official reference
implementation, and it seems like the obvious choice for this information to be
specified.

Section 4.2 of the TUF specification, titled "File formats: general principles"
may be replaced by a description of the properties that any signature wrapper used
by a TUF implementation must provide. Some important properties:

* SHOULD include an authenticated payload type
* SHOULD avoid depending on canonicalization for security
* SHOULD NOT require the verifier to parse the payload before verifying
* SHOULD NOT require the inclusion of signing key algorithms in the signature
* MUST support the inclusion of multiple signatures in a file
* SHOULD support a hint indicating what signing key was used, i.e., a KEYID

The presence of an authenticated payload type can be valuable for a project like TUF,
with multiple implementations and derivatives. Indeed, every POUF that describes an
implementation MUST choose a unique payload type, ensuring that there is no confusion

> about which implementation generated some piece of metadata.

Reviewer: Should the POUF format be updated to include capturing the unique payload type for an implementation? Should we provide any guidance on forming payload type?

The spec should probably more strongly recommend capturing the implementation details in a POUF, especially the payload type.

Author: > Should the POUF format be updated to include capturing the unique payload type for an implementation?

This is probably a good idea. @mnm678 what do you think?

TAP-11 lists only the minimum fields a POUF must contain, so in theory it can be extended without changing the TAP, but it's probably worth adding a field. I also wonder if it's worth adding something to formally identify the implementations a POUF describes. POUF-1 identifies the python implementation, but does it also describe the Go implementation, for example? Should it link to them?

Author: > Should we provide any guidance on forming payload type?

I'm not sure we can, since we're only saying the envelope should support an authenticated payload type, while each envelope format may enforce its own standard. Guidance here may run afoul of some signature wrapper that is otherwise compliant? I think the selection of a unique payload type that conforms to a particular wrapper's specification is sufficient, as long as it's recorded in the corresponding POUF.

Author: Maybe it's worth referring to a survey of wrappers to understand their guidelines for payload types. We may be able to specify some aspects like "include information about encoding" without running into trouble.

Reviewer 2: > > Should the POUF format be updated to include capturing the unique payload type for an implementation?
> 
> This is probably a good idea. @mnm678 what do you think?
> 
> TAP-11 lists only the minimum fields a POUF must contain, so in theory it can be extended without changing the TAP, but it's probably worth adding a field. I also wonder if it's worth adding something to formally identify the implementations a POUF describes. POUF-1 identifies the python implementation, but does it also describe the Go implementation, for example? Should it link to them?

I agree, I think we could add payload type as a subsection of Formats in the POUF definition.

I would say the implementations should/could link to the POUF rather than the other way around. The implementation implements TUF plus some POUF.

Author: > I would say the implementations should/could link to the POUF rather than the other way around. The implementation implements TUF plus some POUF.

Makes sense to me.

Reviewer: SGTM, I think we can file an issue against TAP 11 to discuss payload type once we merge this as draft.

* Author: Aditya Sirish A Yelgundhalli
* Type: Standardization
* Status: Draft
* Content-Type: markdown
* Created: 30/04/2021
* +TUF-Version:
* +Post-History:

# Abstract

This TUF Augmentation Proposal (TAP) proposes switching to a new signature
envelope, namely the new
[SSL signature specification](http://github.com/secure-systems-lab/signing-spec).
This has the following benefits over the current state:

1. Avoids canonicalization for security reasons (i.e., to not parse untrusted
   input)

2. Reduces the possibility of misinterpretation of the payload. The serialized
   payload is encoded as a string and verified by the recipient before
   de-serializing.

The old signature envelope would still be supported but deprecated, possibly to
be removed in a future release. This TAP follows a near identical proposal in
TUF's sister project, in-toto. The proposal there is detailed in an in-toto
Enhancement (ITE),
[ITE-5](https://github.com/in-toto/ITE/blob/master/ITE/5/README.adoc).

# Specification

The specification adopted will be the SSL Signing Spec 0.1, as linked above. As
such, we defer to that document to describe the specifics of signature
generation and verification.

> The envelope's `payloadType` is `application/vnd.tuf+json` for all TUF roles.

Reviewer 2: To include support for TUF implementations that don't use json, can you make this a bit more general? Maybe `application/vnd.tuf+EXT`.

This means that the payload is expected to be a JSON file with a
`_type` field identifying the specific role.

The envelope's `payload` is the JSON serialization of the message, equivalent to
the `signed` object in the current format.

## Pseudocode

Implementations should process the authentication layer as follows:

Inputs:

*   `file`: JSON-encoded TUF document
*   `recognizedSigners`: collection of (`name`, `publicKey`) pairs

Outputs:

*   `message`: the signed message as an implementation-language-specific object
*   `signers`: set of recognized names that have signed the message

Steps:

*   `envelope` := JsonDecode(`file`); raise error if the decoding fails
*   If `envelope.payload` exists (new-style envelope):
    *  If `payloadType` != `application/vnd.tuf+json`, raise error
    *  `preauthEncoding` := PAE(UTF8(`envelope.payloadType`),

>         `envelope.payload`) as per signing-spec

Reviewer 2: Can you link to this part of the spec?

    *  `signers` := set of `name` for which Verify(`preauthEncoding`,
        `signature.sig`, `publicKey`) succeeds, for all combinations of
        (`signature`) in `envelope.signatures` and (`name`, `publicKey`) in
        `recognizedSigners`
    *  `message` := JsonDecode(`envelope.payload`)
*   Else if `envelope.signed` exists (old-style envelope):
    *  `preauthEncoding` := CanonicalJsonEncode(`envelope.signed`)
    *  `signers` := set of `name` for which Verify(`preauthEncoding`,
        `signature.sig`, `publicKey`) succeeds, for all combinations of
        (`signature`) in `envelope.signatures` and (`name`, `publicKey`) in
        `recognizedSigners`
    *  `message` := `envelope.signed`
*   Else, raise error
*   Raise error if `signers` is empty
*   Return `message` and `signers`

# Motivation

> TUF's sister project, in-toto, reused the current signature envelope to maximize

Reviewer 2: ```suggestion
TUF's sister project, in-toto, reused TUF's current signature envelope to maximize
```

[PASETO](https://github.com/paragonie/paseto/blob/master/docs/01-Protocol-Versions/Version2.md#sig)
are overly specific. (Both are also JSON-specific.) We believe the SSL signing
spec strikes the right balance of simplicity, usefulness, and security.

Further, the SSL signing spec is a "natural evolution" of the current signature
envelope, which was defined in both the TUF and in-toto specifications. As such,
it allows for a transparent upgrade via their cryptographic provider,
[securesystemslib](https://github.com/secure-systems-lab/securesystemslib).

Further information on the reasoning behind the envelope's specifics is provided
in the [signing specification](https://github.com/secure-systems-lab/signing-spec)
repository.

# Security Analysis

At first sight this proposal is central to security, yet the actual
contribution is to allow for a signature provider to be disaggregated from the
specification. As such, no software update-specific security properties are
removed from the system through this TAP.

The adoption of SSL signing spec slightly improves the security stance of
implementations because they are no longer parsing untrusted input.

> # Backwards Compatibility

Reviewer 2: We should probably also make a plan to phase out the old signature envelope.

Author: Resolving this, I think that's now part of the POUF.

* TAP: 17
* Title: Remove Signature Wrapper from the TUF Specification

> * Version: 0

Reviewer: ```suggestion
* Version: 1
```

* Last-Modified: 22/06/2021
* Author: Aditya Sirish A Yelgundhalli, Marina Moore
* Type: Standardization
* Status: Draft
* Content-Type: markdown
* Created: 30/04/2021

> * +TUF-Version:

Reviewer: ```suggestion
* Requires: TAP 14
* +TUF-Version:
```
I think this makes sense, because we won't be merging any changes to the spec related to this TAP until TAP 14 has been accepted.

* +Post-History:

# Abstract

This TUF Augmentation Proposal (TAP) proposes removing the definition of a
specific signature wrapper and key definitions, and instead defines certain
properties a wrapper must have. Further, it suggests POUF-1 as an example
implementors can refer to when choosing to generate TUF metadata.

# Specification

> The TUF specification currently, as of June 2021, uses a custom signature

Reviewer: ```suggestion
The TUF specification currently, as of June 2021, uses a custom signature
```
I wonder if we should mention versions, rather than dates, here? TUF uses semver, so we should be able to reference the major version.

wrapper. At the time of authoring this document, the primary reference
implementation written in Python also generates TUF metadata using the same
signature wrapper.

However, TUF does not mandate the use of this signature wrapper, nor any

> specific metaformat. Indeed, TAP-11, "Using POUFs for Interoperability" enables

Reviewer: Nit: link to TAP 11 here might be nice

adopters to make their own decisions for their implementations, and provides a

> mechanism for them to document their decisions. POUF-1 is the POUF for the

Reviewer: Nit: link to POUF 1 here might be nice

official reference implementation, and it seems like the obvious choice for this
information to be specified.

Section 4.2 of the TUF specification, titled "File formats: general principles"
may be replaced by a description of the properties that any signature wrapper used
by a TUF implementation must provide. Some important properties:

> * MUST include an authenticated payload type

Author: Does this list of properties make sense from the perspective of TUF?

* SHOULD avoid depending on canonicalization for security
* SHOULD NOT require the verifier to parse the payload before verifying
* SHOULD NOT require the inclusion of signing key algorithms in the signature
* MUST support the inclusion of multiple signatures in a file
* SHOULD support a hint indicating what signing key was used, i.e., a KEYID

> The presence of an authenticated payload type can be valuable for a project like TUF,

Author: I'd like to hear some thoughts here. Should two implementations designed by the same team to be compatible, say a Python implementation and a Go implementation, have different payload types? Or should they have the same one because in essence they're described by the same POUF? Here I'm leaning towards the latter.

Reviewer 2: I think implementations with the same POUF should have the same payload type. That way it can be used to determine if the metadata is compatible with a particular implementation.

Author: Perhaps the proposed payload type for POUF-1 should contain the POUF number somewhere.

Reviewer: Agree, we want the payload type to describe the format not the implementation. If go-tuf and python-tuf want to have compatible implementations, they should use the same payload type.

with multiple implementations and derivatives. Indeed, every POUF that describes an
implementation MUST choose a unique payload type, ensuring that there is no confusion
about which implementation generated some piece of metadata.

# Motivation

TAP-11 introduced the concept of POUFs but the TUF specification continues to
specify example formats, namely those used by the reference implementation as
of June 2021. These definitions are essentially replicated in POUF-1, which is
meant to be the authoritative source for information about the reference
implementation. By replacing these definitions with *properties* that a wrapper
must possess, the specification can aid adopters with the development of their
implementations and the POUF can serve as an example. In this scenario, both
documents are serving the purpose originally envisioned for them.

Further, the examples used in the specification are from the old signature
wrapper that includes certain side effects:
* it requires canonicalization before signature verification

> * it does not allow for distinguishing different implementations that may have slightly different formats, i.e., it's missing a payload type

Reviewer: Nit: could you wrap this line?

# Rationale

Moving the signature wrapper details out of the specification, and instead

> requiring adopters to refer to POUFs for examples ensures a clean separation

Reviewer: ```suggestion
requiring adopters to refer to POUFs for specifics of an implementation ensures a clean separation
```

between implementation details and the TUF specification. Indeed, it also
ensures that the focus of the reader is on only the TUF primitives rather
than burdening them with the specifics of the signature wrapper.

# Security Analysis

Any implementations that build on the properties listed in this document
will have their security enhanced.

# Backwards Compatibility

> No backwards incompatibility.

Reviewer: I think this section would benefit from some more content. It's strictly true that it's not backwards incompatible for the specification, but the specification has a symbiotic relationship with the reference implementation. A change of the signature wrapper in the reference implementation absolutely is backwards incompatible, as the POUF update indicates.

Reviewer 3: Yeah, we are quite interested in switching to DSSE, and we’d appreciate any guidance on making the transition. Our situation is complicated by us having to support products that don’t support DSSE need to use TUF to update to the latest version.

I’m guessing we will have to either use a tombstone repo which updates us to some new version that points at the new DSSE repo, or we start publishing the metadata in both formats, where the metadata files have different file extensions.

Reviewer 2: @erickt I wrote a [TAP](https://github.com/theupdateframework/taps/blob/master/tap14.md) about managing TUF versions that describes a version of publishing metadata using multiple formats that you might be interested in. I'm hoping we can get that or something similar working before adding this and other breaking changes to the spec.

Author: I just pushed some changes discussing a transition period (on a per implementation basis) during which metadata is published in multiple formats. Also links to TAP-14.

# Augmented Reference Implementation

> See POUF-1 for details about the reference implementation.

Reviewer: Let's mark this TODO?

Reviewer 4: Yeah, where are the POUFs anyway? Are they shared between TUF and Uptane?
